File: output/java/tests/20_transition_pop.java
import java.util.*;


import java.util.*;

class TransitionPopTestFrameEvent {
    String _message;
    HashMap<String, Object> _parameters;

    TransitionPopTestFrameEvent(String message) {
        this._message = message;
        this._parameters = null;
    }

    TransitionPopTestFrameEvent(String message, HashMap<String, Object> parameters) {
        this._message = message;
        this._parameters = parameters;
    }
}

class TransitionPopTestFrameContext {
    TransitionPopTestFrameEvent _event;
    Object _return;
    HashMap<String, Object> _data;

    TransitionPopTestFrameContext(TransitionPopTestFrameEvent event, Object defaultReturn) {
        this._event = event;
        this._return = defaultReturn;
        this._data = new HashMap<>();
    }
}

class TransitionPopTestCompartment {
    String state;
    HashMap<String, Object> state_args;
    HashMap<String, Object> state_vars;
    HashMap<String, Object> enter_args;
    HashMap<String, Object> exit_args;
    TransitionPopTestFrameEvent forward_event;
    TransitionPopTestCompartment parent_compartment;

    TransitionPopTestCompartment(String state) {
        this.state = state;
        this.state_args = new HashMap<>();
        this.state_vars = new HashMap<>();
        this.enter_args = new HashMap<>();
        this.exit_args = new HashMap<>();
        this.forward_event = null;
        this.parent_compartment = null;
    }

    TransitionPopTestCompartment copy() {
        TransitionPopTestCompartment c = new TransitionPopTestCompartment(this.state);
        c.state_args = new HashMap<>(this.state_args);
        c.state_vars = new HashMap<>(this.state_vars);
        c.enter_args = new HashMap<>(this.enter_args);
        c.exit_args = new HashMap<>(this.exit_args);
        c.forward_event = this.forward_event;
        c.parent_compartment = this.parent_compartment;
        return c;
    }
}

class TransitionPopTest {
    private ArrayList<TransitionPopTestCompartment> _state_stack;
    private TransitionPopTestCompartment __compartment;
    private TransitionPopTestCompartment __next_compartment;
    private ArrayList<TransitionPopTestFrameContext> _context_stack;
    public ArrayList log = new ArrayList();

    public TransitionPopTest() {
        _state_stack = new ArrayList<>();
        _context_stack = new ArrayList<>();
        __compartment = new TransitionPopTestCompartment("Idle");
        __next_compartment = null;
        TransitionPopTestFrameEvent __frame_event = new TransitionPopTestFrameEvent("$>");
        TransitionPopTestFrameContext __ctx = new TransitionPopTestFrameContext(__frame_event, null);
        _context_stack.add(__ctx);
        __kernel(_context_stack.get(_context_stack.size() - 1)._event);
        _context_stack.remove(_context_stack.size() - 1);
    }

    private void __kernel(TransitionPopTestFrameEvent __e) {
        __router(__e);
        while (__next_compartment != null) {
            var next_compartment = __next_compartment;
            __next_compartment = null;
            TransitionPopTestFrameEvent exit_event = new TransitionPopTestFrameEvent("<$");
            __router(exit_event);
            __compartment = next_compartment;
            if (__compartment.forward_event == null) {
                TransitionPopTestFrameEvent enter_event = new TransitionPopTestFrameEvent("$>");
                __router(enter_event);
            } else {
                var forward_event = __compartment.forward_event;
                __compartment.forward_event = null;
                if (forward_event._message.equals("$>")) {
                    __router(forward_event);
                } else {
                    TransitionPopTestFrameEvent enter_event = new TransitionPopTestFrameEvent("$>");
                    __router(enter_event);
                    __router(forward_event);
                }
            }
        }
    }

    private void __router(TransitionPopTestFrameEvent __e) {
        String state_name = __compartment.state;
        if (state_name.equals("Idle")) {
            _state_Idle(__e);
        } else if (state_name.equals("Working")) {
            _state_Working(__e);
        }
    }

    private void __transition(TransitionPopTestCompartment next) {
        __next_compartment = next;
    }

    public void start() {
        TransitionPopTestFrameEvent __e = new TransitionPopTestFrameEvent("start");
        TransitionPopTestFrameContext __ctx = new TransitionPopTestFrameContext(__e, null);
        _context_stack.add(__ctx);
        __kernel(_context_stack.get(_context_stack.size() - 1)._event);
        _context_stack.remove(_context_stack.size() - 1);
    }

    public void process() {
        TransitionPopTestFrameEvent __e = new TransitionPopTestFrameEvent("process");
        TransitionPopTestFrameContext __ctx = new TransitionPopTestFrameContext(__e, null);
        _context_stack.add(__ctx);
        __kernel(_context_stack.get(_context_stack.size() - 1)._event);
        _context_stack.remove(_context_stack.size() - 1);
    }

    public String get_state() {
        TransitionPopTestFrameEvent __e = new TransitionPopTestFrameEvent("get_state");
        TransitionPopTestFrameContext __ctx = new TransitionPopTestFrameContext(__e, null);
        _context_stack.add(__ctx);
        __kernel(_context_stack.get(_context_stack.size() - 1)._event);
        var __result = (String) _context_stack.get(_context_stack.size() - 1)._return;
        _context_stack.remove(_context_stack.size() - 1);
        return __result;
    }

    public ArrayList get_log() {
        TransitionPopTestFrameEvent __e = new TransitionPopTestFrameEvent("get_log");
        TransitionPopTestFrameContext __ctx = new TransitionPopTestFrameContext(__e, null);
        _context_stack.add(__ctx);
        __kernel(_context_stack.get(_context_stack.size() - 1)._event);
        var __result = (ArrayList) _context_stack.get(_context_stack.size() - 1)._return;
        _context_stack.remove(_context_stack.size() - 1);
        return __result;
    }

    private void _state_Idle(TransitionPopTestFrameEvent __e) {
        if (__e._message.equals("get_log")) {
            _context_stack.get(_context_stack.size() - 1)._return = this.log;
            return;
        } else if (__e._message.equals("get_state")) {
            _context_stack.get(_context_stack.size() - 1)._return = "Idle";
            return;
        } else if (__e._message.equals("process")) {
            this.log.add("idle:process");
        } else if (__e._message.equals("start")) {
            this.log.add("idle:start:push");
            _state_stack.add(__compartment.copy());
            var __compartment = new TransitionPopTestCompartment("Working");
            __compartment.parent_compartment = this.__compartment.copy();
            __transition(__compartment);
            return;
        }
    }

    private void _state_Working(TransitionPopTestFrameEvent __e) {
        if (__e._message.equals("get_log")) {
            _context_stack.get(_context_stack.size() - 1)._return = this.log;
            return;
        } else if (__e._message.equals("get_state")) {
            _context_stack.get(_context_stack.size() - 1)._return = "Working";
            return;
        } else if (__e._message.equals("process")) {
            this.log.add("working:process:before_pop");
            var __popped = _state_stack.remove(_state_stack.size() - 1);
            __transition(__popped);
            return;
            // This should NOT execute because pop transitions away
            this.log.add("working:process:after_pop");
        }
    }
}

class Main {
    public static void main(String[] args) {
        System.out.println("=== Test 20: Transition Pop (Java) ===");
        var s = new TransitionPopTest();

        // Initial state should be Idle
        if (!s.get_state().equals("Idle")) {
            System.out.println("FAIL: Expected 'Idle', got '" + s.get_state() + "'");
            System.exit(1);
        }
        System.out.println("Initial state: " + s.get_state());

        // start() pushes Idle, transitions to Working
        s.start();
        if (!s.get_state().equals("Working")) {
            System.out.println("FAIL: Expected 'Working', got '" + s.get_state() + "'");
            System.exit(1);
        }
        System.out.println("After start(): " + s.get_state());

        // process() in Working does pop transition back to Idle
        s.process();
        if (!s.get_state().equals("Idle")) {
            System.out.println("FAIL: Expected 'Idle' after pop, got '" + s.get_state() + "'");
            System.exit(1);
        }
        System.out.println("After process() with pop: " + s.get_state());

        ArrayList log = s.get_log();
        System.out.println("Log: " + log);

        // Verify log contents
        if (!log.contains("idle:start:push")) {
            System.out.println("FAIL: Expected 'idle:start:push' in log");
            System.exit(1);
        }
        if (!log.contains("working:process:before_pop")) {
            System.out.println("FAIL: Expected 'working:process:before_pop' in log");
            System.exit(1);
        }
        if (log.contains("working:process:after_pop")) {
            System.out.println("FAIL: Should NOT have 'working:process:after_pop' in log");
            System.exit(1);
        }

        System.out.println("PASS: Transition pop works correctly");
    }
}

File: output/go/tests/20_transition_pop.go

package main

import "fmt"
import "os"

type TransitionPopTestFrameEvent struct {
    _message    string
    _parameters map[string]any
}

type TransitionPopTestFrameContext struct {
    _event  TransitionPopTestFrameEvent
    _return any
    _data   map[string]any
}

type TransitionPopTestCompartment struct {
    state            string
    stateArgs        map[string]any
    stateVars        map[string]any
    enterArgs        map[string]any
    exitArgs         map[string]any
    forwardEvent     *TransitionPopTestFrameEvent
    parentCompartment *TransitionPopTestCompartment
}

func newTransitionPopTestCompartment(state string) *TransitionPopTestCompartment {
    return &TransitionPopTestCompartment{
        state:    state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
    }
}

func (c *TransitionPopTestCompartment) copy() *TransitionPopTestCompartment {
    nc := &TransitionPopTestCompartment{
        state: c.state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
        forwardEvent:     c.forwardEvent,
        parentCompartment: c.parentCompartment,
    }
    for k, v := range c.stateArgs { nc.stateArgs[k] = v }
    for k, v := range c.stateVars { nc.stateVars[k] = v }
    for k, v := range c.enterArgs { nc.enterArgs[k] = v }
    for k, v := range c.exitArgs { nc.exitArgs[k] = v }
    return nc
}

type TransitionPopTest struct {
    _state_stack []*TransitionPopTestCompartment
    __compartment *TransitionPopTestCompartment
    __next_compartment *TransitionPopTestCompartment
    _context_stack []TransitionPopTestFrameContext
    log []string
}

func NewTransitionPopTest() *TransitionPopTest {
    s := &TransitionPopTest{}
    s._state_stack = make([]*TransitionPopTestCompartment, 0)
    s._context_stack = make([]TransitionPopTestFrameContext, 0)
    s.__compartment = newTransitionPopTestCompartment("Idle")
    s.__next_compartment = nil
    __frame_event := TransitionPopTestFrameEvent{_message: "$>", _parameters: nil}
    __ctx := TransitionPopTestFrameContext{_event: __frame_event, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return s
}

func (s *TransitionPopTest) __kernel(__e *TransitionPopTestFrameEvent) {
    s.__router(__e)
    for s.__next_compartment != nil {
        next_compartment := s.__next_compartment
        s.__next_compartment = nil
        exit_event := &TransitionPopTestFrameEvent{_message: "<$", _parameters: s.__compartment.exitArgs}
        s.__router(exit_event)
        s.__compartment = next_compartment
        if s.__compartment.forwardEvent == nil {
            enter_event := &TransitionPopTestFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
            s.__router(enter_event)
        } else {
            forward_event := s.__compartment.forwardEvent
            s.__compartment.forwardEvent = nil
            if forward_event._message == "$>" {
                s.__router(forward_event)
            } else {
                enter_event := &TransitionPopTestFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
                s.__router(enter_event)
                s.__router(forward_event)
            }
        }
    }
}

func (s *TransitionPopTest) __router(__e *TransitionPopTestFrameEvent) {
    switch s.__compartment.state {
    case "Idle":
        s._state_Idle(__e)
    case "Working":
        s._state_Working(__e)
    }
}

func (s *TransitionPopTest) __transition(next *TransitionPopTestCompartment) {
    s.__next_compartment = next
}

func (s *TransitionPopTest) Start() {
    __e := TransitionPopTestFrameEvent{_message: "Start"}
    __ctx := TransitionPopTestFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
}

func (s *TransitionPopTest) Process() {
    __e := TransitionPopTestFrameEvent{_message: "Process"}
    __ctx := TransitionPopTestFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
}

func (s *TransitionPopTest) GetState() string {
    __e := TransitionPopTestFrameEvent{_message: "GetState"}
    __ctx := TransitionPopTestFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result string
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.(string) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *TransitionPopTest) GetLog() []string {
    __e := TransitionPopTestFrameEvent{_message: "GetLog"}
    __ctx := TransitionPopTestFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result []string
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.([]string) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *TransitionPopTest) _state_Working(__e *TransitionPopTestFrameEvent) {
    if __e._message == "GetLog" {
        s._context_stack[len(s._context_stack)-1]._return = s.log
        return
    } else if __e._message == "GetState" {
        s._context_stack[len(s._context_stack)-1]._return = "Working"
        return
    } else if __e._message == "Process" {
        s.log = append(s.log, "working:process:before_pop")
        __popped := s._state_stack[len(s._state_stack)-1]
        s._state_stack = s._state_stack[:len(s._state_stack)-1]
        s.__transition(__popped)
        return
        // This should NOT execute because pop transitions away
        s.log = append(s.log, "working:process:after_pop")
    }
}

func (s *TransitionPopTest) _state_Idle(__e *TransitionPopTestFrameEvent) {
    if __e._message == "GetLog" {
        s._context_stack[len(s._context_stack)-1]._return = s.log
        return
    } else if __e._message == "GetState" {
        s._context_stack[len(s._context_stack)-1]._return = "Idle"
        return
    } else if __e._message == "Process" {
        s.log = append(s.log, "idle:process")
    } else if __e._message == "Start" {
        s.log = append(s.log, "idle:start:push")
        s._state_stack = append(s._state_stack, s.__compartment.copy())
        __compartment := newTransitionPopTestCompartment("Working")
        __compartment.parentCompartment = s.__compartment.copy()
        s.__transition(__compartment)
        return
    }
}

func contains(slice []string, item string) bool {
	for _, v := range slice {
		if v == item {
			return true
		}
	}
	return false
}

func main() {
	fmt.Println("=== Test 20: Transition Pop (Go) ===")
	sm := NewTransitionPopTest()

	// Initial state should be Idle
	if sm.GetState() != "Idle" {
		fmt.Printf("FAIL: Expected 'Idle', got '%s'\n", sm.GetState())
		os.Exit(1)
	}
	fmt.Printf("Initial state: %s\n", sm.GetState())

	// start() pushes Idle, transitions to Working
	sm.Start()
	if sm.GetState() != "Working" {
		fmt.Printf("FAIL: Expected 'Working', got '%s'\n", sm.GetState())
		os.Exit(1)
	}
	fmt.Printf("After Start(): %s\n", sm.GetState())

	// process() in Working does pop transition back to Idle
	sm.Process()
	if sm.GetState() != "Idle" {
		fmt.Printf("FAIL: Expected 'Idle' after pop, got '%s'\n", sm.GetState())
		os.Exit(1)
	}
	fmt.Printf("After Process() with pop: %s\n", sm.GetState())

	log := sm.GetLog()
	fmt.Printf("Log: %v\n", log)

	// Verify log contents
	if !contains(log, "idle:start:push") {
		fmt.Println("FAIL: Expected 'idle:start:push' in log")
		os.Exit(1)
	}
	if !contains(log, "working:process:before_pop") {
		fmt.Println("FAIL: Expected 'working:process:before_pop' in log")
		os.Exit(1)
	}
	if contains(log, "working:process:after_pop") {
		fmt.Println("FAIL: Should NOT have 'working:process:after_pop' in log")
		os.Exit(1)
	}

	fmt.Println("PASS: Transition pop works correctly")
}
